File: src/main/java/com/pauljoda/modularsystems/power/tiles/TileBankSolids.java
package com.pauljoda.modularsystems.power.tiles;

import com.pauljoda.modularsystems.power.container.ContainerSolidsBank;
import com.pauljoda.modularsystems.power.gui.GuiSolidsBank;
import com.teambr.bookshelf.collections.InventoryTile;
import com.teambr.bookshelf.common.tiles.IOpensGui;
import com.teambr.bookshelf.helpers.GuiHelper;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.ISidedInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntityFurnace;
import net.minecraft.world.World;

import java.util.List;

public class TileBankSolids extends TileBankBase implements IOpensGui, ISidedInventory {

    private InventoryTile inventory;
    private int cooldown;

    public TileBankSolids() {
        inventory = new InventoryTile(27);
        cooldown = 0;
    }

    @Override
    public double getPowerLevelScaled(int scale) {
        return getFuelCount() * scale / inventory.getSizeInventory();
    }

    /**
     * Used to count how many slots have things in them
     * @return How many slots have an item
     */
    private int getFuelCount() {
        int count = 0;
        for (int i = 0; i < inventory.getSizeInventory(); i++) {
            if (inventory.getStackInSlot(i) != null && inventory.getStackInSlot(i).stackSize > 0)
                count++;
        }
        return count;
    }

    /**
     * Helper Method to consume solid fuels
     */
    private int consumeFuel(boolean simulate) {
        for (int i = 0; i < inventory.getSizeInventory(); i++) {
            if (inventory.getStackInSlot(i) != null && inventory.getStackInSlot(i).stackSize > 0 &&
                    TileEntityFurnace.isItemFuel(inventory.getStackInSlot(i))) {
                int burnValue = TileEntityFurnace.getItemBurnTime(inventory.getStackInSlot(i));
                if (!simulate) {
                    decrStackSize(i, 1);
                    worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
                }
                return burnValue;
            }
        }
        return 0;
    }

    /*******************************************************************************************************************
     ***************************************** Fuel Provider Functions *************************************************
     *******************************************************************************************************************/


    @Override
    public boolean canProvide() {
        return consumeFuel(true) > 0;
    }

    @Override
    public double fuelProvided() {
        return consumeFuel(true);
    }

    @Override
    public double consume() {
        return consumeFuel(false);
    }

    @Override
    public FuelProviderType type() {
        return FuelProviderType.ITEM;
    }

    /*******************************************************************************************************************
     *************************************** Inventory Methods *********************************************************
     *******************************************************************************************************************/

    @Override
    public int getSizeInventory() {
        return inventory.getSizeInventory();
    }

    @Override
    public ItemStack getStackInSlot(int slot) {
        return inventory.getStackInSlot(slot);
    }

    @Override
    public ItemStack decrStackSize(int slot, int count) {
        ItemStack itemStack = inventory.getStackInSlot(slot);
        if (itemStack != null) {
            if (itemStack.stackSize <= count) {
                setInventorySlotContents(slot, null);
            }
            itemStack = itemStack.splitStack(count);
        }
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
        return itemStack;
    }

    @Override
    public ItemStack getStackInSlotOnClosing(int slot) {
        ItemStack stack = getStackInSlot(slot);
        if (stack != null) {
            setInventorySlotContents(slot, null);
        }
        return stack;
    }

    @Override
    public void setInventorySlotContents(int slot, ItemStack stack) {
        inventory.setStackInSlot(stack, slot);
    }

    @Override
    public int getInventoryStackLimit() {
        return 64;
    }

    @Override
    public boolean isItemValidForSlot(int i, ItemStack itemstack) {
        return TileEntityFurnace.isItemFuel(itemstack);
    }

    @Override
    public boolean canInsertItem(int i, ItemStack itemstack, int j) {
        return getCore() != null && isItemValidForSlot(i, itemstack);
    }

    @Override
    public boolean canExtractItem(int i, ItemStack itemstack, int j) {
        return false;
    }

    @Override
    public int[] getAccessibleSlotsFromSide(int var1) {
        int[] sides = new int[inventory.getSizeInventory()];
        for (int x = 0; x < inventory.getSizeInventory(); x++) {
            sides[x] = x;
        }
        return sides;
    }

    /*******************************************************************************************************************
     ******************************************** Tile Methods *********************************************************
     *******************************************************************************************************************/

    @Override
    public void readFromNBT (NBTTagCompound tags)
    {
        super.readFromNBT(tags);
        inventory.readFromNBT(tags);
        cooldown = tags.getInteger("cooldown");
    }

    @Override
    public void writeToNBT (NBTTagCompound tags)
    {
        super.writeToNBT(tags);
        inventory.writeToNBT(tags);
        tags.setInteger("cooldown", cooldown);
    }

    /*******************************************************************************************************************
     ****************************************** IOpensGui Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public Object getServerGuiElement(int i, EntityPlayer entityPlayer, World world, int i1, int i2, int i3) {
        return new ContainerSolidsBank(entityPlayer.inventory, this);
    }

    @Override
    public Object getClientGuiElement(int i, EntityPlayer entityPlayer, World world, int i1, int i2, int i3) {
        return new GuiSolidsBank(entityPlayer.inventory, this);
    }

    /*******************************************************************************************************************
     ************************************************ Waila ************************************************************
     *******************************************************************************************************************/

    @Override
    public void returnWailaHead(List<String> list) {
        list.add(GuiHelper.GuiColor.YELLOW + "Filled Fuel Slots: " + GuiHelper.GuiColor.WHITE + getFuelCount() + "/" + inventory.getSizeInventory());
    }

    @Override
    public void returnWailaTail(List<String> list) {
        if (Minecraft.getMinecraft().thePlayer.isSneaking()) {
            list.add(GuiHelper.GuiColor.CYAN + GuiHelper.GuiTextFormat.ITALICS.toString() + "Useable In:");
            list.add(GuiHelper.GuiColor.GREEN + "Modular Furnace");
            list.add(GuiHelper.GuiColor.GREEN + "Modular Crusher");
            list.add(GuiHelper.GuiColor.GREEN + "Modular Generator");
        } else
            list.add(GuiHelper.GuiColor.CYAN + GuiHelper.GuiTextFormat.ITALICS.toString() + "Press Shift for Usage Cores");

        super.returnWailaTail(list);
    }
}

File: src/main/scala/com/teambr/modularsystems/power/tiles/TileBankSolids.scala
package com.teambr.modularsystems.power.tiles

import com.teambr.bookshelf.common.tiles.traits.Inventory
import com.teambr.modularsystems.core.providers.FuelProvider.FuelProviderType
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntityFurnace
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.items.CapabilityItemHandler

/**
  * This file was created for Modular-Systems
  *
  * Modular-Systems is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since August 07, 2015
  */
class TileBankSolids extends TileBankBase with Inventory {

    /**
      * Used to scale the current power level
      *
      * @param scale The scale to move to
      * @return A number from 0 - { @param scale} for current level
      */
    override def getPowerLevelScaled(scale: Int): Double = getFuelCount * scale / 27

    /**
      * Used to count how many slots have things in them
      *
      * @return How many slots have an item
      */
    private def getFuelCount: Int = {
        var count = 0
        for (i <- 0 until getSizeInventory) {
            if (getStackInSlot(i) != null && getStackInSlot(i).stackSize > 0)
                count += 1
        }
        count
    }

    /**
      * Helper Method to consume solid fuels
      */
    private def consumeFuel(simulate: Boolean): Int = {
        for (i<- 0 until getSizeInventory) {
            if (getStackInSlot(i) != null && getStackInSlot(i).stackSize > 0 && TileEntityFurnace.isItemFuel(getStackInSlot(i))) {
                val burnValue = TileEntityFurnace.getItemBurnTime(getStackInSlot(i))
                if (!simulate) {
                    decrStackSize(i, 1)
                    worldObj.notifyBlockUpdate(pos, worldObj.getBlockState(pos), worldObj.getBlockState(pos), 6)
                }
                return burnValue
            }
        }
        0
    }

    /*******************************************************************************************************************
      **************************************** Fuel Provider Functions *************************************************
      ******************************************************************************************************************/
    override def canProvide: Boolean = consumeFuel(true) > 0

    override def fuelProvided(): Double = consumeFuel(true)

    override def consume(): Double = consumeFuel(false)

    override def `type`(): FuelProviderType = FuelProviderType.ITEM

    /*******************************************************************************************************************
      ************************************** Inventory Methods *********************************************************
      ******************************************************************************************************************/

    override def initialSize: Int = 27


    override def hasCapability(capability: Capability[_], facing : EnumFacing) = {
        capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY
    }

    override def getCapability[T](capability: Capability[T], facing: EnumFacing): T = {
        super[Inventory].getCapability(capability, facing)
    }

    override def getSizeInventory : Int = {
        super[Inventory].getSizeInventory
    }

    override def getStackInSlot(index : Int) : ItemStack = {
        super[Inventory].getStackInSlot(index)
    }

    override def decrStackSize(index : Int, count : Int) : ItemStack = {
        super[Inventory].decrStackSize(index, count)
    }

    override def setInventorySlotContents(index : Int, stack : ItemStack) : Unit = {
        super[Inventory].setInventorySlotContents(index, stack)
    }

    override def getInventoryStackLimit : Int = {
        super[Inventory].getInventoryStackLimit
    }

    override def isUseableByPlayer(player : EntityPlayer) : Boolean = {
        super[Inventory].isUseableByPlayer(player)
    }

    override def isItemValidForSlot(index : Int, stack : ItemStack) : Boolean = {
        TileEntityFurnace.getItemBurnTime(stack) > 0
    }

    override def clear() : Unit = {
        super[Inventory].clear()
    }

    override def getSlotsForFace(side : EnumFacing) : Array[Int] = {
        val array = new Array[Int](27)
        for(i <- 0 to 27)
            array(i) = i
        array
    }

    override def canExtractItem(index : Int, stack : ItemStack, direction : EnumFacing) : Boolean = {
        true
    }

    override def canInsertItem(index : Int, itemStackIn : ItemStack, direction : EnumFacing) : Boolean = {
        TileEntityFurnace.getItemBurnTime(itemStackIn) > 0
    }

    override def removeStackFromSlot(index: Int): ItemStack = {
        super[Inventory].removeStackFromSlot(index)
    }

    override def readFromNBT(tag: NBTTagCompound) : Unit = {
        super[TileBankBase].readFromNBT(tag)
        super[Inventory].readFromNBT(tag)
    }

    override def writeToNBT(tag: NBTTagCompound) : NBTTagCompound = {
        super[TileBankBase].writeToNBT(tag)
        super[Inventory].writeToNBT(tag)
        tag
    }

    override def markDirty(): Unit = {
        super[TileBankBase].markDirty()
        super[Inventory].markDirty()
    }
}
